### PR TITLE
Add issue 49 reproduction scripts

### LINKED ISSUE #49: 【2026犀牛鸟开源人才专属】【低难度】模型训练专项-垂类数据集基线训练
在 VisDrone（航拍密集小目标）和 SKU-110K（零售密集商品）等垂类数据集上(可选，换用其他的垂类公开数据集)，分别使用 YOLO-Master-v0.1-N 和 YOLO-Master-EsMoE-N 进行训练复现
- 使用项目内置数据集配置 ultralytics/cfg/datasets/VisDrone.yaml 和 ultralytics/cfg/datasets/SKU-110K.yaml
- 训练配置参考论文推荐：imgsz=640，epochs 100~300（根据 GPU 资源调整）
- 记录完整训练日志，包含每 epoch 的 mAP50、mAP50-95、box_loss、cls_loss、moe_loss(使用wandb进行配置，可以将url设置为公开贴出来)
- 在scripts/reproduce/ 下提交可复现脚本 reproduce_visdrone.py 和 reproduce_sku110k.py
提供 README.md 说明数据集下载命令、训练命令、预期结果、已知问题与解决方案
提交 Pull Request，包含训练脚本、运行日志、结果对比表
https://github.com/Tencent/YOLO-Master
https://github.com/Tencent/YOLO-Master/releases/tag/YOLO-Master-v26.02
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/cfg/datasets/VisDrone.yaml
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/cfg/datasets/SKU-110K.yaml
https://github.com/Tencent/YOLO-Master/blob/main/scripts/compare_moe_coco128.py
难度：低

---

**本issue为2026犀牛鸟开源人才培养活动专属issue，仅供已报名参与犀牛鸟活动的同学认领**
【认领时间】7月1日～7月31日（7月1日前认领视为无效❗️）
【认领方式】在本issue评论区回复“已认领本任务”，即视为认领成功
【活动报名】需提前完成犀牛鸟报名问卷，问卷将用于活动登记和奖励发放：https://wj.qq.com/s2/26888567/gh2q

### PR DESCRIPTION
**Pull request by 2021113691**
## Summary

Fixes #49.

This PR adds reproducible scripts and recorded run results for Rhino-Bird issue #49 on VisDrone and SKU-110K. It compares the two official Nano models:

| Model | Config |
| --- | --- |
| YOLO-Master-v0.1-N | `ultralytics/cfg/models/master/v0_1/det/yolo-master-n.yaml` |
| YOLO-Master-EsMoE-N | `ultralytics/cfg/models/master/v0/det/yolo-master-n.yaml` |

After reviewer feedback, I also reran EsMoE-N with the `dense_eval_fix` setting. The previous `pre_fix_sparse` validation path used sparse inference during validation and substantially under-reported EsMoE accuracy. With dense expert aggregation during validation, EsMoE-N aligns with or slightly exceeds YOLO-Master-v0.1-N on both datasets.

## Changes

- Added reproducible scripts:
  - `scripts/reproduce/reproduce_visdrone.py`
  - `scripts/reproduce/reproduce_sku110k.py`
  - `scripts/reproduce/reproduce_common.py`
- Added `scripts/reproduce/README.md` with dataset preparation, training commands, W&B links, result tables, and known issues.
- Added `scripts/reproduce/issue49_results.csv` with compact recorded metrics for PR review.
- Fixed EsMoE eval/validation behavior in `ultralytics/nn/modules/moe/modules.py` so dense expert aggregation is used by default; sparse inference remains explicit opt-in for latency benchmarking.
- Added regression coverage in `tests/test_moe.py` for dense-by-default eval and explicit sparse inference opt-in.
- Added robust dataset-cache handling to avoid corrupted `labels.cache` when concurrent runs prepare the same dataset.
- Added a lightweight `ultralytics.utils.lora.sensitivity` compatibility shim because the current checkout imports this module.

## Results

Final-epoch metrics:

| Model | Dataset | Eval | Epochs | mAP50 | mAP50-95 |
| --- | --- | --- | ---: | ---: | ---: |
| YOLO-Master-v0.1-N | VisDrone | default | 100 | 0.29555 | 0.17046 |
| YOLO-Master-EsMoE-N | VisDrone | pre_fix_sparse | 100 | 0.08387 | 0.04225 |
| YOLO-Master-EsMoE-N | VisDrone | dense_eval_fix | 100 | 0.30612 | 0.17615 |
| YOLO-Master-v0.1-N | SKU-110K | default | 100 | 0.89253 | 0.56774 |
| YOLO-Master-EsMoE-N | SKU-110K | pre_fix_sparse | 54 | 0.09703 | 0.04207 |
| YOLO-Master-EsMoE-N | SKU-110K | dense_eval_fix | 100 | 0.89498 | 0.57162 |

Full recorded metrics, including best-checkpoint epochs, losses, and W&B run URLs, are available in:

`[scripts/reproduce/issue49_results.csv](scripts/reproduce/issue49_results.csv)`

## W&B

Project:

https://wandb.ai/2063055270-harbin-institute-of-technology/yolo-master-issue-49

Key runs:

- VisDrone v0.1-N: https://wandb.ai/2063055270-harbin-institute-of-technology/yolo-master-issue-49/runs/5fi8avsl
- VisDrone EsMoE-N `dense_eval_fix`: https://wandb.ai/2063055270-harbin-institute-of-technology/yolo-master-issue-49/runs/4v15e3o8
- SKU-110K v0.1-N: https://wandb.ai/2063055270-harbin-institute-of-technology/yolo-master-issue-49/runs/6np8a4lu
- SKU-110K EsMoE-N `dense_eval_fix`: https://wandb.ai/2063055270-harbin-institute-of-technology/yolo-master-issue-49/runs/oiomzvte

## Notes

The original EsMoE validation path used sparse inference, which prunes low-confidence experts and can under-report validation accuracy. The `dense_eval_fix` reruns use dense expert aggregation for validation consistency while keeping sparse inference available as an explicit opt-in path for latency benchmarking.

SKU-110K contains several JPEG warnings such as `Corrupt JPEG data`; training and validation continue normally. One W&B online upload attempt timed out while logging plots, so the fixed SKU-110K EsMoE run was recorded offline and synced after completion.

## Verification

- Reproduced 100-epoch VisDrone runs.
- Reproduced 100-epoch SKU-110K runs.
- Synced W&B logs for final runs.
- Added test coverage for EsMoE dense eval and explicit sparse inference opt-in.
- Verified no dataset files, run outputs, W&B local files, or checkpoints are committed.
- Ran `python -m py_compile scripts/reproduce/reproduce_common.py scripts/reproduce/reproduce_visdrone.py scripts/reproduce/reproduce_sku110k.py`.
- Ran `python scripts/reproduce/reproduce_visdrone.py --dry-run`.
- Ran `git diff --check`.

### CODE REVIEW

**Comment by skywalker-lt** [comment]
The YOLO-Master-EsMoE-N model didn't validate correctly on both datasets because you might have used `sparse_foward` during inference. I think you might need to amend the inference logic. If the validation is performed correctly, the EsMoE-M model should perfom identical to the v0.1-N, like:

```bash
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
    213/300        63G      1.295     0.8829     0.9369  2.079e-05        816        640: 100% ━━━━━━━━━━━━ 102/102 3.3it/s 31.3s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.2it/s 2.2s
                   all        548      38759      0.445      0.346      0.345      0.201

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
    214/300        63G      1.289     0.8774     0.9357   2.03e-05        561        640: 100% ━━━━━━━━━━━━ 102/102 3.3it/s 31.2s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
                   all        548      38759      0.448      0.344      0.345      0.201

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
    215/300        63G      1.287     0.8768     0.9341  1.968e-05        512        640: 100% ━━━━━━━━━━━━ 102/102 3.2it/s 31.9s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
                   all        548      38759      0.448      0.344      0.345      0.201

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
    216/300        63G      1.286     0.8704     0.9346  1.891e-05        689        640: 100% ━━━━━━━━━━━━ 102/102 3.2it/s 31.6s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
                   all        548      38759      0.449      0.344      0.345      0.201
```

**Comment by 2021113691** [comment]
> The YOLO-Master-EsMoE-N model didn't validate correctly on both datasets because you might have used `sparse_foward` during inference. I think you might need to amend the inference logic. If the validation is performed correctly, the EsMoE-M model should perfom identical to the v0.1-N, like:
> 
> ```shell
>       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
>     213/300        63G      1.295     0.8829     0.9369  2.079e-05        816        640: 100% ━━━━━━━━━━━━ 102/102 3.3it/s 31.3s
>                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.2it/s 2.2s
>                    all        548      38759      0.445      0.346      0.345      0.201
> 
>       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
>     214/300        63G      1.289     0.8774     0.9357   2.03e-05        561        640: 100% ━━━━━━━━━━━━ 102/102 3.3it/s 31.2s
>                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
>                    all        548      38759      0.448      0.344      0.345      0.201
> 
>       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
>     215/300        63G      1.287     0.8768     0.9341  1.968e-05        512        640: 100% ━━━━━━━━━━━━ 102/102 3.2it/s 31.9s
>                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
>                    all        548      38759      0.448      0.344      0.345      0.201
> 
>       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss   moe_loss  Instances       Size
>     216/300        63G      1.286     0.8704     0.9346  1.891e-05        689        640: 100% ━━━━━━━━━━━━ 102/102 3.2it/s 31.6s
>                  Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5/5 2.4it/s 2.1s
>                    all        548      38759      0.449      0.344      0.345      0.201
> ```
Thanks a lot for pointing this out. You are right: the previous EsMoE validation could run through the sparse inference path, which makes the eval behavior inconsistent with the dense expert aggregation used during training and may under-report the validation metrics.

I have started fixing this by changing `ES_MOE` to use dense expert aggregation by default during eval/validation, while keeping sparse inference as an explicit opt-in path for latency benchmarking. I also added/updated a test to cover this behavior.

I am rerunning the EsMoE experiments on VisDrone and SKU-110K with the fixed validation logic and `patience=0`, and will update the PR results/W&B links once the new runs finish.
